intake_esm/core.py

```diff
@@ -58,31 +57,171 @@ class esm_datastore(intake.catalog.Catalog):
+            aggregation_dict = {}
+            variable_column_name = self.esmcol_data['aggregation_control']['variable_column_name']
+            groupby_attrs = self.esmcol_data['aggregation_control'].get('groupby_attrs', [])
+            aggregations = self.esmcol_data['aggregation_control'].get('aggregations', [])
+            # Sort aggregations to make sure join_existing is always done before join_new
+            aggregations = sorted(aggregations, key=lambda i: i['type'], reverse=True)
+            for agg in aggregations:
+                key = agg['attribute_name']
+                rest = agg.copy()
+                del rest['attribute_name']
+                aggregation_dict[key] = rest
+            agg_columns = list(aggregation_dict.keys())
+
+        if not groupby_attrs:
+            groupby_attrs = self.df.columns.tolist()
+
+        # filter groupby_attrs to ensure no columns with all nans
+        def _allnan_or_nonan(column):
+            if self.df[column].isnull().all():
+                return False
+            elif self.df[column].isnull().any():
+                raise ValueError(
+                    f'The data in the {column} column should either be all NaN or there should be no NaNs'
+                )
+            else:
+                return True
+
+        groupby_attrs = list(filter(_allnan_or_nonan, groupby_attrs))
+
+        info = {
+            'groupby_attrs': groupby_attrs,
+            'variable_column_name': variable_column_name,
+            'aggregations': aggregations,
+            'agg_columns': agg_columns,
+            'aggregation_dict': aggregation_dict,
+            'path_column_name': path_column_name,
+            'data_format': data_format,
+            'format_column_name': format_column_name,
+        }
+        return info
+
+    def keys(self):
+        keys = list(map(lambda x: self.sep.join(x), self._keys))
+        return keys
+
+    @property
+    def key_template(self):
+        return self.sep.join(self.aggregation_info['groupby_attrs'])
+
+    def __len__(self):
+        return len(self.keys())
+
+    def _get_entries(self):
+        # Due to just-in-time entry creation, we may not have all entries loaded
+        # We need to make sure to create entries missing from self._entries
+        if len(self._entries) != len(self.keys()):
+            for key in self.keys():
+                if key not in self._entries:
+                    self[key]
+        return self._entries
+
+    def __getitem__(self, key):
+        # The canonical unique key is the key of a compatible group of assets
+        try:
+            return self._entries[key]
+        except KeyError:
+            if key in self.keys():
+                _key = tuple(key.split(self.sep))
+                df = self._grouped.get_group(_key)
+                self._entries[key] = _make_entry(key, df, self.aggregation_info)
+                return self._entries[key]
+            else:
+                raise KeyError(key)
+
```

Comment:
@jhamman, I like this implementation of `__getitem__()`. It clearly draws a line between searching for an item, and getting an item. Thank you for the idea!